pkg/loki/modules.go

```diff
@@ -296,6 +297,13 @@ func (t *Loki) initQueryFrontend() (_ services.Service, err error) {
 	frontend.RegisterFrontendServer(t.server.GRPC, t.frontend)
 
 	frontendHandler := queryrange.StatsHTTPMiddleware.Wrap(t.httpAuthMiddleware.Wrap(t.frontend.Handler()))
+
+	httpMiddleware := middleware.Merge(
+		t.httpAuthMiddleware,
+		queryrange.StatsHTTPMiddleware,
+	)
+	tp := tailproxy.New(t.cfg.Frontend, util.Logger)
+	wsHandler := httpMiddleware.Wrap(http.HandlerFunc(tp.Handle))
```

Comment:
I think we do not need all the code for managing the whole lifecycle of WebSocket. I think we should use httputil.ReverseProxy which would reduce the code to just following:
```
	tp := httputil.NewSingleHostReverseProxy(t.cfg.Frontend.DownstreamURL)
	wsHandler := httpMiddleware.Wrap(tp)
```
I am sorry you have put in a lot of efforts to write all the code but I think we should use that battle-tested code.

Also, I do not think we should be using `DownstreamURL` because it is meant to be used for roundtripping the requests directly to downstream server instead of queuing them to be picked up by grpc client running on queries. We should have a separate config for getting the URL for proxying WebSocket requests.

Reply:
Thanks for the review. I will test is with the `NewSingleHostReverseProxy`. I thought that it is only for HTTP requests. At least I get more familiar with web sockets ;)

I will also add separate config option for the WS downstream.